src/main/java/com/github/pgasync/PgConnection.java

```diff
@@ -0,0 +1,362 @@
+import com.github.pgasync.message.frontend.Parse;
+import com.github.pgasync.message.frontend.PasswordMessage;
+import com.github.pgasync.message.frontend.Query;
+import com.github.pgasync.message.frontend.StartupMessage;
+
+/**
+ * A connection to Postgres backend. The postmaster forks a backend process for
+ * each connection. A connection can process only a single query at a time.
+ *
+ * @author Antti Laisi
+ */
+public class PgConnection implements Connection {
+
+    /**
+     * Uses named server side prepared statement and named portal.
+     */
+    public class PgPreparedStatement implements PreparedStatement {
+
+        private final String sname;
+        private Columns columns;
+
+        PgPreparedStatement(String sname) {
+            this.sname = sname;
+        }
+
+        @Override
+        public CompletableFuture<ResultSet> query(Object... params) {
+            List<Row> rows = new ArrayList<>();
+            return fetch((columnsByName, orderedColumns) -> {
+            }, rows::add, params)
+                    .thenApply(v -> new PgResultSet(columns.byName, List.of(columns.ordered), rows, 0));
+        }
+
+        @Override
+        public CompletableFuture<Integer> fetch(BiConsumer<Map<String, PgColumn>, PgColumn[]> onColumns, Consumer<Row> processor, Object... params) {
+            Bind bind = new Bind(sname, dataConverter.fromParameters(params));
+            Consumer<DataRow> rowProcessor = dataRow -> processor.accept(new PgRow(dataRow, columns.byName, columns.ordered, dataConverter));
+            if (columns != null) {
+                onColumns.accept(columns.byName, columns.ordered);
+                return stream
+                        .send(bind, rowProcessor);
+            } else {
+                return stream
+                        .send(bind, Describe.portal(), columnDescriptions -> {
+                            columns = calcColumns(columnDescriptions);
+                            onColumns.accept(columns.byName, columns.ordered);
+                        }, rowProcessor);
+            }
+        }
+
+        @Override
+        public CompletableFuture<Void> close() {
+            return stream.send(Close.statement(sname)).thenAccept(closeComplete -> {
+            });
+        }
+    }
+
+    public static class Columns {
+        final Map<String, PgColumn> byName;
+        final PgColumn[] ordered;
+
+        Columns(Map<String, PgColumn> byName, PgColumn[] ordered) {
+            this.byName = byName;
+            this.ordered = ordered;
+        }
+    }
+
+    private static class NameSequence {
+
+        private long counter;
+        private String prefix;
+
+        NameSequence(final String prefix) {
+            this.prefix = prefix;
+        }
+
+        private String next() {
+            if (counter == Long.MAX_VALUE) {
+                counter = 0;
+                prefix = "_" + prefix;
+            }
+            return prefix + ++counter;
```

Comment:
Not thread-safe, counter should be an AtomicLong.

Reply:
Yes, currently it is a bug. But there is another option. According to Postgres protocol, prepared statements exist within a Postgres session only, and so wouldn't be better to refactor the only use case of `NameSequence` and make it instance non static? This will help us to avoid counter contention.

Reply:
Yeah, that sounds like a better choice.

Reply:
Fixed.